### ISSUE TITLE
Plans to implement ieee 754-2008 optional operations

### ISSUE THREAD

**Issue by keyneom** [CraneStation/cranelift#653]
Would operations such as `pow` be expected to be added at some point? Which recommended operations from clause 9 of ieee 754-2008 will be added?

**Comment by sunfishcode** [comment]
IEEE 754 is typically not implented purely in hardware or programming language operators; most implementations make use of libraries to implent much of the required and recommend functionality. Cranelift is similar; the expected way to do a `pow` is to call `pow`, from libm.

**Comment by keyneom** [comment]
Thank you, in that case, how would you call `pow` using the `FunctionBuilder` and two IR `Value`s?

**Comment by sunfishcode** [comment]
You'd call `import_function` and import a function that the linking phase can resolve to "pow". How to do that varies depending on how you're running the code. Is this for a JIT, or a fairie .o file? Are using cranelift-module, cranelift-simplejit, and/or cranelift-faerie?

**Comment by keyneom** [comment]
I'm using cranelift-module and cranelift-simplejit and trying to implement something similar to [CraneStation/simplejit-demo](https://github.com/CraneStation/simplejit-demo).

**Comment by keyneom** [comment]
I think I follow, it's a bit unclear to me how
> The puts function is resolved by dlsym to the libc function
but I think I can spend sometime looking into what that means and how to leverage it for my own functions.

Thank you for all the help! It's my first time attempting to generate IR, so I'm still trying to figure things out.

**Comment by sunfishcode** [comment]
Sounds good. Feel free to ask if anything is unclear!

**Comment by sunfishcode** [comment]
One way to do this is to explicitly register functions you want to call with [symbol](https://docs.rs/cranelift-simplejit/0.26.0/cranelift_simplejit/struct.SimpleJITBuilder.html#method.symbol) -- you give it a name and a raw function pointer to one of your rust functions, and the JIT will thereafter resolve that name to that pointer.

That's not super safe though :-). Actually, it should probably be explicitly `unsafe`, because it can lead to the JIT calling an arbitrary function pointer.

So yeah, it would be good to have a safer interface here. I don't have a full plan here, but I think one key thing is to make sure that we have a proper function declaration registered with `Module::declare_function` that we can associate a Rust function with. We can either have an interface where you pass in the `FuncId`, or one where it looks up the `FuncId` based on the name. Then we'll also want to pass in a full Rust function reference, rather than a raw pointer. If we can then check that the function reference's type matches the `ir::Signature` of the declaration, then in theory that could be a safe interface. Does that make sense?

**Comment by keyneom** [comment]
Thanks I'll give that a try!

In terms of a safer means of implementing the function method, I was thinking that it might be possible to create a procedural macro for a derive trait that would convert a fn definition to a static struct of some kind that would copy the fn definition as an impl for the struct. Then when wanting to call the fn you would execute it using an `invoke` method of sorts on the struct. The struct would, upon creation, take care of exporting itself and making itself available for dlsym. The struct would expose the required methods (or properties) to either call itself as a direct method or to do so through dlsym.

Did any of that make sense? Would it be useful in providing a safer interface? Let me know your thoughts and I might have some time to try putting something together. We've got a new baby here at home right now so I can't make a ton of guarantees on timelines or anything, but if you feel like it might be useful I can try to work it in!

**Comment by sunfishcode** [comment]
Yes, a procedural macro like what you describe to convert from a static Rust type into a dynamic description of a callable function would be very useful in creating a safe interface. I myself am not very familiar with procedural macros yet, so I don't know exactly how this should look, but if you wanted to investigate them, I can help if you have any questions about what we need them to do.

Take care with the new baby!
